content/docs/ai/ai-checkpoints.md

```diff
@@ -0,0 +1,356 @@
+
+    // Terminal states - safe to proceed
+    if (['finished', 'skipped', 'cancelled'].includes(status)) return
+
+    // Error state - handle appropriately
+    if (status === 'failed') throw error
+
+    sleep(5 seconds)
+  }
+}
+
+// After restore API call
+operationIds = restoreResponse.operations.map(op => op.id)
+if (operationIds.length > 0) {
+  operationIds.forEach(id => waitForOperation(id))
+}
+// NOW safe to connect
+```
+
+<Admonition type="tip" title="Working example">
+See `lib/neon/apply-snapshot.ts` and `lib/neon/operations.ts` in the demonstration application at https://github.com/neondatabase-labs/snapshots-as-checkpoints-demo
+</Admonition>
+
+<Admonition type="note" title="todo">
+**Visual Diagram: Operations Polling Flow** - Flowchart showing restore API call, operations returned, polling loop with state transitions (running → finished/failed), danger zones where connection attempts fail, and the final "safe to connect" state
+</Admonition>
+
+**Common restore errors:**
+
+- **Connection to old state**: Ensure all operations completed
+- **Target branch not found**: Verify branch exists
+- **Operation timeout**: Retry with longer timeout
+- **Accumulating backup branches**: Delete old backup branches (e.g., `production (old)`) after successful restore verification
+
+#### Preview environments
+
+Create a temporary branch to preview a checkpoint without affecting the active branch:
+
+```json
+{
+  "name": "preview-checkpoint-123",
+  "finalize_restore": false // Creates new branch for preview without moving computes
+}
+```
+
+This creates a new branch with its own connection string for preview. The active branch remains unchanged.
+
+### Managing checkpoints
+
+#### List available checkpoints
+
+Get all snapshots with IDs, names, and timestamps using the [list snapshots endpoint](https://api-docs.neon.tech/reference/listsnapshots):
+
+```bash
+GET /api/v2/projects/{project_id}/snapshots
+```
+
+**Path Parameters:**
+
+- `project_id` (string, required): The Neon project ID
+
+#### Delete checkpoint
+
+Remove a snapshot using the [delete endpoint](https://api-docs.neon.tech/reference/deletesnapshot):
+
+```bash
+DELETE /api/v2/projects/{project_id}/snapshots/{snapshot_id}
+```
+
+**Path parameters:**
+
+- `project_id` (string, required): The Neon project ID
+- `snapshot_id` (string, required): The snapshot ID
+
+#### Update checkpoint name
+
+Rename a snapshot using the [update endpoint](https://api-docs.neon.tech/reference/updatesnapshot):
+
+```bash
+PATCH /api/v2/projects/{project_id}/snapshots/{snapshot_id}
+```
+
+**Path parameters:**
+
+- `project_id` (string, required): The Neon project ID
+- `snapshot_id` (string, required): The snapshot ID
+
+**Body:**
+
+```json
+{
+  "snapshot": {
+    "name": "important-milestone"
+  }
+}
+```
+
+#### Cleanup strategy
+
+Proper cleanup reduces costs and keeps your project manageable:
```

Comment:
Cleanup is a key item in this use case. Can we provide an example?
We also need to see where we're at with the automatic deletion (ttl) for snapshots. We can ask Alec in the project snapshots channel.